Implement AXExpanded handling for command buttons

lukewarlow opened the pull request:
#### 362e1d2e9abae57701df04c5bee563bcbb4be833
<pre>
Implement AXExpanded handling for command buttons
<a href="https://bugs.webkit.org/show_bug.cgi?id=269672">https://bugs.webkit.org/show_bug.cgi?id=269672</a>

Reviewed by Tyler Wilcock and Darin Adler.

This patch updates WebKit to handle exposing the expanded state on command buttons.

Specifically when the command button is linked with a popover,
and has a valid command for popover it will expose the expanded state.

See <a href="https://github.com/w3c/aria/pull/2354">https://github.com/w3c/aria/pull/2354</a>

* LayoutTests/accessibility/mac/expanded-notification-commandfor-popover-expected.txt: Added.
* LayoutTests/accessibility/mac/expanded-notification-commandfor-popover.html: Added.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::updateIsolatedTree):
(WebCore::AXObjectCache::relationAttributes):
(WebCore::AXObjectCache::attributeToRelationType):
(WebCore::AXObjectCache::updateRelations):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::commandForElement const):
(WebCore::AccessibilityNodeObject::commandType const):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::commandType const):
(WebCore::AccessibilityObject::supportsExpanded const):
(WebCore::AccessibilityObject::isExpanded const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::commandForElement const):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::popoverState const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/html/HTMLButtonElement.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::popoverState const): Deleted.
* Source/WebCore/html/HTMLElement.h:

Canonical link: <a href="https://commits.webkit.org/292067@main">https://commits.webkit.org/292067@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c8549b740a43a60f07223bca4ad11759559bd3db

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94808 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14399 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4236 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99824 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45297 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96855 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14674 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22816 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72317 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29620 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97809 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10940 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85605 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52649 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10637 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3320 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44637 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80862 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3419 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101867 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21836 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15950 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81316 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22088 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81624 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80696 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20171 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25264 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2664 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/15069 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21816 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26933 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21476 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24947 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23217 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

keithamus commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3453,7 +3458,12 @@ bool AccessibilityObject::isExpanded() const
+            if (invokeTargetElement->isPopoverShowing())
+                return true;
+            if (RefPtr dialog = dynamicDowncast<HTMLDialogElement>(invokeTargetElement))
+                return dialog->isOpen();
```
```suggestion
```

We likely don't want expanded on Dialog invokers as they're only going to show as modal, so the information about the expanded state is a little redundant (you know if a modal dialog is expanded because you'll be in it).

lukewarlow commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3395,6 +3406,11 @@ bool AccessibilityObject::isExpanded() const
 
     if (supportsExpanded()) {
+        if (RefPtr commandForElement = this->commandForElement()) {
+            if (RefPtr dialogElement = dynamicDowncast<HTMLDialogElement>(commandForElement); dialogElement && commandType() == CommandType::Close)
```
If #41288 gets merged first this will need to account for that too (and also the test will need updating)

lukewarlow replied:
No longer applicable

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3353,8 +3358,11 @@ bool AccessibilityObject::supportsPressed() const
 {
-    // If this object can toggle an HTML popover, it supports the reporting of its expanded state (which is based on the expanded / collapsed state of that popover).
-    if (popoverTargetElement())
+    // Command buttons support expanded, if they can show, hide or toggle a popover.
```
Slight rephrasing that might help this read a bit better:

`// Elements with commands should support the expanded state if they can show, hide, or toggle a popover.`

IMO it's OK to not specifically call out buttons in the comment, since there's technically no code local to this function that is button-specific (even though the implementation of `commandForElement` only works for buttons).

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -2811,7 +2811,9 @@ void AXObjectCache::handleAttributeChange(Element* element, const QualifiedName&
         recomputeParentTableProperties(element, TableProperty::CellSlots);
-    } else if (attrName == popovertargetAttr)
+    } else if (attrName == commandforAttr)
+        postNotification(element, AXNotification::CommandForChanged);
```
Don't we also need to listen for `commandAttr` changes? e.g. if it's removed, or added, the given element's potential to support expanded changes, right?

lukewarlow replied:
Ah that's a good point, I'll update to account for that and add to the test to check that behaviour.

twilco reviewed (commented):
You may already be aware of this, but WebKit accessibility has two "modes". EWS runs the first — all accessibility requests are served from the main-thread, using the `AccessibilityObject` class hierarchy, and therefore the DOM and render tree.

The second is called "isolated tree mode", which currently [runs in a post-commit bot](https://build.webkit.org/#/builders/1231). This serves accessibility requests from a cached copy of the accessibility tree on a secondary thread. It's important your test passes in this configuration too. You can pass the `--accessibility-isolated-tree` flag to `Tools/Scripts/run-webkit-tests` to run tests in this mode.

darinadler commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -5069,6 +5077,7 @@ Vector<QualifiedName>& AXObjectCache::relationAttributes()
         aria_labeledbyAttr,
         aria_ownsAttr,
         headersAttr,
+        commandforAttr,
```
I think we want this sorted alphabetically, so `commandfor` would come before `headers`.

darinadler commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -5123,7 +5132,7 @@ AXRelationType AXObjectCache::attributeToRelationType(const QualifiedName& attri
     if (attribute == aria_activedescendantAttr)
         return AXRelationType::ActiveDescendant;
-    if (attribute == aria_controlsAttr || attribute == popovertargetAttr)
+    if (attribute == aria_controlsAttr || attribute == popovertargetAttr || attribute == commandforAttr)
```
Even here we like to keep things sorted so the order is not arbtrary:
```suggestion
   if (attribute == aria_controlsAttr || attribute == commandforAttr || attribute == popovertargetAttr)
```

darinadler commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -166,6 +166,8 @@ class AccessibilityReplacedText {
     macro(ColumnSpanChanged) \
     macro(ContentEditableAttributeChanged) \
+    macro(CommandChanged) \
+    macro(CommandForChanged) \
```
Here too we like to sort:
```suggestion
    macro(CommandChanged) \
    macro(CommandForChanged) \
    macro(ContentEditableAttributeChanged) \
```

darinadler commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1177,6 +1178,18 @@ RefPtr<Element> AccessibilityNodeObject::popoverTargetElement() const
 
+RefPtr<Element> AccessibilityNodeObject::commandForElement() const
+{
+    WeakPtr buttonElement = dynamicDowncast<HTMLButtonElement>(node());
```
The local should be `RefPtr`, not `WeakPtr`. I don’t know why other code in this file is using `WeakPtr` for local variables, but it’s not a good pattern either for efficiency or correctness.
```suggestion
    RefPtr buttonElement = dynamicDowncast<HTMLButtonElement>(node());
```
I also would have named the local variable just `element`, even though it’s always a button element. The single word names are easier to read, and it’s not like we rely on the name to understand the type here.

darinadler commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -1177,6 +1178,18 @@ RefPtr<Element> AccessibilityNodeObject::popoverTargetElement() const
+
+CommandType AccessibilityNodeObject::commandType() const
+{
+    WeakPtr buttonElement = dynamicDowncast<HTMLButtonElement>(node());
```
The local should be `RefPtr`, not `WeakPtr`. I don’t know why other code in this file is using `WeakPtr` for local variables, but it’s not a good pattern either for efficiency or correctness.
```suggestion
    RefPtr buttonElement = dynamicDowncast<HTMLButtonElement>(node());
```
I also would have named the local variable just `element`, even though it’s always a button element. The single word names are easier to read, and it’s not like we rely on the name to understand the type here.

darinadler commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3353,8 +3358,11 @@ bool AccessibilityObject::supportsPressed() const
+    // Elements with commands should support the expanded state if they can show, hide, or toggle a popover.
+    if (RefPtr targetElement = commandForElement(); targetElement && targetElement->popoverState() != PopoverState::None) {
+        if (commandType() == CommandType::ShowPopover || commandType() == CommandType::HidePopover || commandType() == CommandType::TogglePopover)
+            return true;
```
Is there some way to write this command type check so it’s not just a list of 3 types? I worry that having this code in accessibility doesn’t work well if we later refactor and add new command types; we won’t get a warning if we add a new type that affects popovers. I’d rather have a named function which expresses what these 3 command types have in common that uses a switch statement. Not sure what the word for that is, maybe "affects popover visibility", or some better phrase. Maybe local to this file, or even maybe in the DOM itself rather than in accessibility code, for easier maintenance.

I’m also wondering about the difference between the popover state check on the target element, which does not skip over the popover target element check below, and the command type check on this object, which does skip over it. Is it important to have these two be different? Maybe there could be a function that combines both checks? Currently the code seems a little arbitrary and I think a named function might help the reader understand both of these small issues.

lukewarlow replied:
I've swapped this over to using a switch and added some comments in, eefactored it a bit too. I've not moved it to a separate function. This is because I don't want to design this in such a way that it gets unrefactored the moment we need to do AX handling for other commands. It's currently popover specific but we'll likely need it for future commands.

darinadler commented on Source/WebCore/accessibility/AccessibilityObject.h:
```diff
@@ -442,6 +444,8 @@ class AccessibilityObject : public AXCoreObject, public CanMakeWeakPtr<Accessibi
     virtual Element* anchorElement() const { return nullptr; }
     virtual RefPtr<Element> popoverTargetElement() const { return nullptr; }
+    virtual RefPtr<Element> commandForElement() const { return nullptr; }
+    virtual CommandType commandType() const;
```
Maybe we should expose a property more specific to how this is used, like a boolean predicate related to popovers or related to whether we should export expanded state for popovers, rather than `CommandType`? Also, I think this can be `private` for now rather than `public`. It’s only used inside `AccessibilityObject` itself.

lukewarlow replied:
I've made this private, similar to above I don't want to make this too popover specific so have left it as commandType().

darinadler commented on Source/WebCore/accessibility/AccessibilityObject.h:
```diff
@@ -442,6 +444,8 @@ class AccessibilityObject : public AXCoreObject, public CanMakeWeakPtr<Accessibi
     static AccessibilityObject* headingElementForNode(Node*);
     virtual Element* anchorElement() const { return nullptr; }
     virtual RefPtr<Element> popoverTargetElement() const { return nullptr; }
+    virtual RefPtr<Element> commandForElement() const { return nullptr; }
```
I know this name is matching a function in the DOM, but it’s a peculiar name for a function. Can we give it a more straightforward name?

lukewarlow replied:
I'm somewhat biased from the fact this is an API I've worked with quite heavily but commandForElement seems as simple a name as possible for this function? It's a function that returns the value of the commandForElement property (which in turn reflects the element referenced in the commandfor attribute). The same way that popoverTargetElement (the function above this) works.

darinadler replied:
Oh, I see, it’s the element that the command is for? I do not like the name “command for element” for that. I’d call it “element for command”, in fact!

lukewarlow replied:
It's the element pointed to by the commandfor attribute. This returns an element rather than a command (which is a separate attribute).

commandfor is akin to popovertarget, command is akin to popovertargetaction.

darinadler replied:
What can I say, "command for element" is so deeply confusing when it doesn’t mean "the command for an element" but rather "the element pointed to by the commandfor attribute". Deeply mysterious design!

annevk replied:
HTML attempts to consistently use an Element suffix for Element-reflecting IDL attributes. I do agree that it becomes a bit confusing here due to the attribute name ending in "for" and we could decide to rename it for the non-DOM cases, but then we should probably also rename popoverTargetElement?

darinadler commented on Source/WebCore/dom/Element.h:
```diff
@@ -102,6 +102,11 @@ enum class FullscreenNavigationUI : uint8_t;
+    None,
+    Auto,
+    Manual,
+};
```
I think the single-line horizontal format is nicer for small enumerations like this. Like `IsSyntheticClick` above and `ResolveURLs` below.

darinadler commented on Source/WebCore/dom/Element.h:
```diff
@@ -675,6 +680,7 @@ class Element : public ContainerNode {
     PopoverData& ensurePopoverData();
     void clearPopoverData();
     bool isPopoverShowing() const;
+    PopoverState popoverState() const;
```
I know we are just moving this and hoisting it from button element, but does the term “popover state” come from a specification? I ask because it’s vague and, for example, it seems that “whether the popover is showing“ is also “popover state”. This is something more specific than that. Not sure what the best words would be for it. If it does come from how it’s described in the specification, then I think it’s fine to stay with this terminology.

I am not fond with how the code related to popovers is spread across `Element`, `HTMLElement`, `HTMLFormControlElement`, and `HTMLButtonElement`. I wonder if there’s a way for us to make this less sprawling. I’d want to move more of it down to `HTMLElement`, but this instead moves more of it up to `Element`; it will only ever be set by `HTMLElement::popoverAttributeChanged`, and I think we should move the member functions for it out of `Element`.

lukewarlow replied:
PopoverState represents this table from the spec. I guess PopoverAttributeState might be more specific a name? https://html.spec.whatwg.org/multipage/popover.html#attr-popover-auto

Please could you file this as an issue to discuss and address in a follow-up? I'd prefer not to refactor popover more widely than needed for what is quite a scoped patch.

darinadler replied:
I see. It’s an enumeration representing the value of the `popover` attribute, but as an enumeration. I agree that `PopoverAttributeValue` or `PopoverAttributeValue` are other possible names that are less confusing than `PopoverState` but calling things “attribute value” or “attribute state” is irritatingly wordy. I could imagine calling this “popover behavior”, because this particular “state” specifies how the popover should behave, and not the state of the popover.

Maybe a better design would be to not return these values (and looks like we don’t implement hint yet!), but instead consider having getters that return the answers to questions like "should close other popovers when opened" and "should light dismiss" and "should respond to close requests".

Anyway, I don’t insist that you fix this all in this refactoring. But I’m not thrilled by the clarity of this code we’re modifying here.

darinadler reviewed (approved):
This looks fine as is, but there is some refactoring here that doesn’t seem as good as it could be.

lukewarlow committed:
Implement AXExpanded handling for command buttons
https://bugs.webkit.org/show_bug.cgi?id=269672

Reviewed by Tyler Wilcock and Darin Adler.

This patch updates WebKit to handle exposing the expanded state on command buttons.

Specifically when the command button is linked with a popover,
and has a valid command for popover it will expose the expanded state.

See https://github.com/w3c/aria/pull/2354

* LayoutTests/accessibility/mac/expanded-notification-commandfor-popover-expected.txt: Added.
* LayoutTests/accessibility/mac/expanded-notification-commandfor-popover.html: Added.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::handleAttributeChange):
(WebCore::AXObjectCache::updateIsolatedTree):
(WebCore::AXObjectCache::relationAttributes):
(WebCore::AXObjectCache::attributeToRelationType):
(WebCore::AXObjectCache::updateRelations):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::commandForElement const):
(WebCore::AccessibilityNodeObject::commandType const):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::commandType const):
(WebCore::AccessibilityObject::supportsExpanded const):
(WebCore::AccessibilityObject::isExpanded const):
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::commandForElement const):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::popoverState const):
* Source/WebCore/dom/Element.h:
* Source/WebCore/html/HTMLButtonElement.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::popoverState const): Deleted.
* Source/WebCore/html/HTMLElement.h:

Canonical link: https://commits.webkit.org/292067@main
